File: shared/nutrition-engine/arabicNormalize.ts
/**
 * منفذ حرفي من arabic_normalize.py (Python) — نفس السلوك بالضبط، بدون أي تعديل بالمنطق.
 * تطبيع النص العربي لتحسين مطابقة أسماء الأكل: يوحّد أشكال الألف، يزيل التشكيل والتطويل،
 * وينظّف المسافات فقط — لا يربط كلمات مختلفة عشوائيًا.
 */

const DIACRITICS = /[ؐ-ًؚ-ٰٟۖ-ۭ]/g;
const TATWEEL = "ـ";

const ALEF_FORMS: Record<string, string> = {
  "أ": "ا", "إ": "ا", "آ": "ا", "ٱ": "ا",
  "ى": "ي", // ألف مقصورة -> ياء (شائع بالكتابة العراقية)
};

function translateAlef(text: string): string {
  let out = "";
  for (const ch of text) {
    out += ALEF_FORMS[ch] ?? ch;
  }
  return out;
}

export function normalize(text: string | null | undefined): string {
  if (!text) return "";
  let t = text.trim();
  t = t.replace(DIACRITICS, "");
  t = t.split(TATWEEL).join("");
  t = translateAlef(t);
  t = t.replace(/\s+/g, " ");
  return t.trim();
}

File: arabic_normalize.py
"""
تطبيع النص العربي لتحسين مطابقة أسماء الأكل — بدون ربط كلمات مختلفة عشوائيًا.
يوحّد أشكال الألف، يزيل التشكيل والتطويل، وينظّف المسافات فقط.
"""
import re

_DIACRITICS = re.compile(r"[\u0610-\u061A\u064B-\u065F\u0670\u06D6-\u06ED]")
_TATWEEL = "\u0640"

_ALEF_FORMS = str.maketrans({
    "أ": "ا", "إ": "ا", "آ": "ا", "ٱ": "ا",
    "ى": "ي",  # ألف مقصورة -> ياء (شائع بالكتابة العراقية)
})


def normalize(text: str) -> str:
    if not text:
        return ""
    text = text.strip()
    text = _DIACRITICS.sub("", text)
    text = text.replace(_TATWEEL, "")
    text = text.translate(_ALEF_FORMS)
    text = re.sub(r"\s+", " ", text)
    return text.strip()
